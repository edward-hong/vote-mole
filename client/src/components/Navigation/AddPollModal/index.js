import React from 'react'
import { connect } from 'react-redux'
import { reduxForm, Field, FieldArray } from 'redux-form'
import { withRouter } from 'react-router-dom'
import {
	Modal,
	ModalHeader,
	ModalBody,
	ModalFooter,
	Button,
	Form,
} from 'reactstrap'
import { ajax } from 'rxjs/ajax'
import { compose, or, map, equals } from 'ramda'

import AddPollQuestionInput from './AddPollQuestionInput'
import AddPollOptionsInputs from './AddPollOptionsInputs'
import BlackButton from '../../styled/BlackButton'

const AddPollModal = ({
	isOpen,
	toggle,
	reset,
	handleSubmit,
	pristine,
	submitting,
	auth,
	history,
}) => {
	const cancel = () => {
		reset()
		toggle()
	}

	// Submit poll to backend
	const submit = ({ pollQuestion, options }) =>
		ajax({
			url: '/poll/submit',
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body: {
				userID: auth._id,
				pollQuestion: pollQuestion,
				pollOptions: map(option => ({ option }), options),
			},
		}).subscribe(
			({ response }) => {
				cancel()
				// history.push(`/poll/${response}`)
			},
			({ status }) => {
				if (equals(status, 406)) {
					toggle()
				}
			}
		)

	return (
		<Modal isOpen={isOpen} toggle={toggle}>
			<ModalHeader toggle={cancel}>Add Poll</ModalHeader>
			<ModalBody>
				<Form onSubmit={handleSubmit(submit)} id="add_poll">
					<Field
						name="pollQuestion"
						type="text"
						label="Poll Question"
						placeholder="Enter your poll question"
						component={AddPollQuestionInput}
					/>
					<FieldArray name="options" component={AddPollOptionsInputs} />
				</Form>
			</ModalBody>
			<ModalFooter>
				<Button outline type="submit" form="add_poll" color="success">
					Submit
				</Button>{' '}
				<Button
					outline
					color="secondary"
					disabled={or(pristine, submitting)}
					onClick={reset}>
					Clear Values
				</Button>
				<BlackButton outline onClick={cancel}>
					Cancel
				</BlackButton>
			</ModalFooter>
		</Modal>
	)
}

// Validate add poll form
const validate = ({ pollQuestion, options }) => {
	const errors = {}

	if (!pollQuestion) errors.pollQuestion = 'Poll question is required'

	if (!options || options.length < 2) {
		errors.options = ['Poll must have at least 2 options']
	} else if (options && options.length > 20) {
		errors.options = ['No more than 20 options allowed']
	} else {
		const optionsArrayErrors = []
		options.forEach((option, i) => {
			if (!option || !option.length) {
				optionsArrayErrors[i] = 'Please fill out all options fields'
			}
		})
		if (optionsArrayErrors.length) {
			errors.options = optionsArrayErrors
		}
	}

	if (options && options[0]) {
		const unique = [...new Set(options)]
		if (unique.length !== options.length) {
			errors.options = ['Options must be unique']
		}
	}

	return errors
}

const mapStateToProps = ({ form, auth }) => ({ form, auth })

export default compose(
	reduxForm({ form: 'add_poll', validate }),
	connect(mapStateToProps),
	withRouter
)(AddPollModal)
