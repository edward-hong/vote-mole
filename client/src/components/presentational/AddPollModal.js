import React from 'react'
import { Modal } from 'reactstrap'

const AddPollModal = ({ isOpen, toggle }) => (
	<Modal isOpen={isOpen} toggle={toggle}>
		AddPollModal
	</Modal>
)

export default AddPollModal
