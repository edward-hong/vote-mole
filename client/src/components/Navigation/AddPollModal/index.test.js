import React from 'react'
import { fireEvent } from '@testing-library/react'
import { from } from 'rxjs'
import { ajax } from 'rxjs/ajax'

import AddPollModal from '.'
import renderWithRouterAndRedux from '../../../utils/test/renderWithRouterAndRedux'

const resolvePromise = new Promise((resolve, reject) => {
	resolve({ response: 'resolved' })
})
const rejectPromise = status =>
	new Promise((resolve, reject) => {
		reject({ status })
	})

jest.mock('rxjs/ajax', () => ({
	ajax: jest.fn(),
}))

const testSubmitPoll = () => {
	const toggleExpandNav = jest.fn()
	const toggle = jest.fn()
	const { getByText, getByPlaceholderText } = renderWithRouterAndRedux(
		<AddPollModal
			isOpen={true}
			toggleExpandNav={toggleExpandNav}
			toggle={toggle}
		/>,
		{ initialState: { form: 'add_poll', auth: false } },
	)
	const addOptionButton = getByText('Add Option')
	const pollQuestionInput = getByPlaceholderText('Enter your poll question')
	const form = document.getElementById('add_poll')

	fireEvent.click(addOptionButton)
	fireEvent.click(addOptionButton)
	fireEvent.change(pollQuestionInput, { target: { value: 'a' } })
	const options1 = document.getElementById('options[0]')
	const options2 = document.getElementById('options[1]')
	fireEvent.change(options1, { target: { value: 'a' } })
	fireEvent.change(options2, { target: { value: 'b' } })
	fireEvent.submit(form)

	expect(ajax).toHaveBeenCalled()
}

ajax.mockImplementationOnce(() => from(resolvePromise))
ajax.mockImplementationOnce(() => from(rejectPromise(406)))
ajax.mockImplementationOnce(() => from(rejectPromise(401)))
ajax.mockImplementationOnce(() => from(rejectPromise(600)))

describe('AddPollModal', () => {
	it('returns null when `isOpen` prop is false', () => {
		const { container } = renderWithRouterAndRedux(
			<AddPollModal isOpen={false} />,
			{ initialState: { form: 'add_poll', auth: false } },
		)

		expect(container.firstChild).toBeNull()
	})

	it('renders correctly', () => {
		const { container } = renderWithRouterAndRedux(
			<AddPollModal isOpen={true} />,
			{ initialState: { form: 'add_poll', auth: false } },
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('will show error if more than 20 options are added on submit', () => {
		const { getByText } = renderWithRouterAndRedux(
			<AddPollModal isOpen={true} />,
			{ initialState: { form: 'add_poll', auth: false } },
		)
		const addOptionButton = getByText('Add Option')
		const form = document.getElementById('add_poll')

		for (let i = 0; i < 21; i++) {
			fireEvent.click(addOptionButton)
		}
		fireEvent.submit(form)

		const errorMessage = getByText('No more than 20 options allowed')
		expect(errorMessage).toBeInTheDocument()
	})

	it('multiple same options will cause an error', () => {
		const { getByText } = renderWithRouterAndRedux(
			<AddPollModal isOpen={true} />,
			{ initialState: { form: 'add_poll', auth: false } },
		)
		const addOptionButton = getByText('Add Option')
		const form = document.getElementById('add_poll')

		fireEvent.click(addOptionButton)
		fireEvent.click(addOptionButton)
		const options1 = document.getElementById('options[0]')
		const options2 = document.getElementById('options[1]')
		fireEvent.change(options1, { target: { value: 'test' } })
		fireEvent.change(options2, { target: { value: 'test' } })
		fireEvent.submit(form)

		const errorMessage = getByText('Options must be unique')
		expect(errorMessage).toBeInTheDocument()
	})

	it('calls the submit handler when form is submitted', testSubmitPoll)

	it(
		'handles error if poll submitted has a duplicated poll question',
		testSubmitPoll,
	)

	it('handles error if user is not logged in to submit poll', testSubmitPoll)

	it('handles error', testSubmitPoll)
})
