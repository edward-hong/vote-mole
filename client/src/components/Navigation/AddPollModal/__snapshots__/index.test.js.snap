// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPollModal renders correctly 1`] = `null`;
